React/Views/RCTView.m

```diff
@@ -172,13 +170,32 @@ - (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
   BOOL isPointInside = [self pointInside:point withEvent:event];
   BOOL needsHitSubview = !(_pointerEvents == RCTPointerEventsNone || _pointerEvents == RCTPointerEventsBoxOnly);
   if (needsHitSubview && (![self clipsToBounds] || isPointInside)) {
+    // Take z-index into account when calculating the touch target.
+    // Check if sorting is required - in most cases it won't be.
+    BOOL sortingRequired = NO;
+    for (UIView *subview in self.subviews) {
+      if (subview.layer.zPosition != 0) {
+        sortingRequired = YES;
+        break;
+      }
+    }
+    NSArray<UIView *> *sortedSubviews = sortingRequired ? [self.reactSubviews sortedArrayUsingComparator:^NSComparisonResult(UIView *a, UIView *b) {
```

Comment:
As you know, `hitTest` must be incredibly performant, so I believe we have to decouple this logic and introduce caching.
I think we can implement at least two cache invalidation strategies:
 * Use weak references and invalidate inside something like `setNeedsLayout`;
 * Use strong NSArray and destroy it in all `insert/add/remove/Subview` methods family (I personally prefer this one even if it requires overriding half of dozen methods).